### PR TITLE
Added validation severity setting for Qute `UndefinedVariable`

### PR DESCRIPTION
**Pull request by AlexXuChen**
Added validation severity setting for `UndefinedVariable` Qute error code.

Part of #451
Requires https://github.com/redhat-developer/vscode-quarkus/pull/446

Signed-off-by: Alexander Chen <alchen@redhat.com>

### CODE REVIEW

File: qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-quickstart/.classpath [hunk 1]
```diff
@@ -1,32 +1,49 @@
...
-	</classpathentry>
-	<classpathentry kind="con" path="org.eclipse.m2e.MAVEN2_CLASSPATH_CONTAINER">
-		<attributes>
-			<attribute name="maven.pomderived" value="true"/>
-		</attributes>
-	</classpathentry>
-	<classpathentry kind="output" path="target/classes"/>
-</classpath>
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-quickstart/.classpath:2]: Dont updta this file

File: qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-quickstart/src/main/resources/templates/ItemResource/items.qute.html [hunk 2]
```diff
@@ -4,14 +4,15 @@
...
+<title>Qute - Items</title>
 </head>
-<body> 
+<body>
     <h1>List of Items</h1>
     <ul>
         {#for item in items}
         <li>
             {item.name}:
+            {item or 'John'}
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-quickstart/src/main/resources/templates/ItemResource/items.qute.html]: Dont update this file

File: qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-quickstart/src/main/resources/templates/ItemResource/items.qute.html [hunk 3]
```diff
@@ -4,14 +4,13 @@
 <html>
 <head>
 <meta charset="UTF-8">
-<title>Qute - Items</title> 
+<title>Qute - Items</title>
```
Comment (angelozerr) [inline qute.jdt/com.redhat.qute.jdt.test/projects/maven/qute-quickstart/src/main/resources/templates/ItemResource/items.qute.html:7]: Dont update this file

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java [hunk 4]
```diff
@@ -173,6 +174,14 @@ public QuteDiagnostics(JavaDataModelCache javaCache) {
 		return diagnostics;
 	}
 
+	private void validateSyntaxProperty(QuteValidationSettings validationSettings) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java:177]: I suppose its just for testing severity?

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationSettings.java [hunk 5]
```diff
@@ -30,13 +38,15 @@
 
 	private boolean enabled;
 
+	private QuteValidationTypeSettings undefined;
+
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationSettings.java:42]: Rename undefined to undefinedVariable

**Review by angelozerr — CHANGES_REQUESTED** [review]

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	public void setSeverity(String severity) {
+		this.severity = severity;
+	}
+
+	/**
+	 * Returns the array of properties to ignore for this validation type.
+	 *
+	 * @return the array of properties to ignore for this validation type.
+	 */
+	public List<String> getExcluded() {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java:53]: remove

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		return excluded;
+	}
+
+	/**
+	 * Set the array of properties to ignore for this validation type.
+	 *
+	 * @param excluded the array of properties to ignore for this validation type.
+	 */
+	public void setExcluded(List<String> excluded) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java:62]: remove

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+	/**
+	 * Returns the diagnostic severity according the given property name and null
+	 * otherwise.
+	 *
+	 * @param propertyName the property name.
+	 * @return the diagnostic severity according the given property name and null
+	 *         otherwise.
+	 */
+	public DiagnosticSeverity getDiagnosticSeverity(String propertyName) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java:74]: remove

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+		return null;
+	}
+
+	/**
+	 * Returns true if the given property name must be excluded and false otherwise.
+	 *
+	 * @param propertyName the property name
+	 * @return true if the given property name must be excluded and false otherwise.
+	 */
+	private boolean isExcluded(String propertyName) {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java:98]: remove

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationSettings.java [hunk 10]
```diff
@@ -30,13 +38,15 @@
 
 	private boolean enabled;
 
+	private QuteValidationTypeSettings undefinedVariable;
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationSettings.java:41]: Replace equals and hashCode with:

	```
@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + (enabled ? 1231 : 1237);
		result = prime * result + ((undefinedVariable == null) ? 0 : undefinedVariable.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		QuteValidationSettings other = (QuteValidationSettings) obj;
		if (enabled != other.enabled)
			return false;
		if (undefinedVariable == null) {
			if (other.undefinedVariable != null)
				return false;
		} else if (!undefinedVariable.equals(other.undefinedVariable))
			return false;
		return true;
	}
```

Please note that this code has been generated by Eclipse IDE "Source / Generate hashCode() and equals()..."

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+		List<String> excludedProperties = getExcluded();
+		for (String excluded : excludedProperties) {
+			// the property name matches an excluded pattern
+			if (excluded.equals(propertyName)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationTypeSettings.java:112]: Add

```java
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((severity == null) ? 0 : severity.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		QuteValidationTypeSettings other = (QuteValidationTypeSettings) obj;
		if (severity == null) {
			if (other.severity != null)
				return false;
		} else if (!severity.equals(other.severity))
			return false;
		return true;
	}

```

**Comment by angelozerr** [comment]
@AlexXuChen here a branch which shouldwork https://github.com/angelozerr/quarkus-ls/tree/539-fix

I suggest that you copy/paste in your PR the Java file from https://github.com/angelozerr/quarkus-ls/blob/539-fix/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteDiagnostics.java

And for me it's working, here a demo:

SeverityDemo

**Comment by angelozerr** [comment]
@AlexXuChen could you please:

 * resolve conflicts
 * rename UndefinedVariable with UndefinedObject for error code and settings please and for the error message to (don't use variable but object)

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationSettings.java [hunk 12]
```diff
@@ -13,15 +13,23 @@
...
  * Qute validation settings.
- * 
+ *
  * @author Angelo ZERR
  *
  */
 public class QuteValidationSettings {
 
+	private static enum Severity {
+		none, error, warning;
```
Comment (fbricon) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/settings/QuteValidationSettings.java:23]: s/none/ignore

**Comment by AlexXuChen** [comment]
Will restart after https://github.com/redhat-developer/quarkus-ls/pull/578 is merged

**Pull request closed without merging**